From 6a122ad4117bedfe5cdf91f6c923172896b4c4af Mon Sep 17 00:00:00 2001
From: Levente Polyak <anthraxx@archlinux.org>
Date: Tue, 16 Sep 2025 17:54:45 +0200
Subject: [PATCH] fix(repo-updated): fix certain repos not being updated due to
 sorting

The GraphQL query is sorting by last_activity_at, but we are iterating
sequentially through updated_at while assuming its in perfect order to
be able to early exit.
The sorting of a different field lead to entries being put in wrong
order which lead to an early exit and some repos being skipped.
Unfortunately the GraphQL API does not allow to sort namespace projects
by updated_at, hence we need to migrate to last_activity_at which has
the side effect its only persistent max once an hour plus contains way
more events compared to the updated_at. Subsequently we will pull more
repos then strictly needed.

As a followup, we should contribute a path or issue to GitLab to be able
to sort by updated_at.

Fixes #180
---
 buildbtw-poc/src/gitlab/gitlab_changed_projects.graphql | 1 +
 buildbtw-poc/src/gitlab/mod.rs                          | 8 ++++----
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/buildbtw-poc/src/gitlab/gitlab_changed_projects.graphql b/buildbtw-poc/src/gitlab/gitlab_changed_projects.graphql
index 1044839..1e633a2 100644
--- a/buildbtw-poc/src/gitlab/gitlab_changed_projects.graphql
+++ b/buildbtw-poc/src/gitlab/gitlab_changed_projects.graphql
@@ -4,6 +4,7 @@ query ChangedProjects($after: String, $group: ID!) {
             nodes {
                 name
                 updatedAt
+                lastActivityAt
             }
             pageInfo {
                 endCursor
diff --git a/buildbtw-poc/src/gitlab/mod.rs b/buildbtw-poc/src/gitlab/mod.rs
index ea8f58a..03b6cd3 100644
--- a/buildbtw-poc/src/gitlab/mod.rs
+++ b/buildbtw-poc/src/gitlab/mod.rs
@@ -32,12 +32,12 @@ pub async fn fetch_all_source_repo_changes(
             result.first()
         );
         last_fetched = first_result
-            .updated_at
+            .last_activity_at
             .clone()
             .map(OffsetDateTime::from)
-            // Work around inaccuracy of the `updated_at` field
+            // Work around inaccuracy of the `updated_at` and `last_activity_at` field
             // https://gitlab.archlinux.org/archlinux/buildbtw/-/issues/32
-            .map(|date| date - Duration::minutes(6));
+            .map(|date| date - Duration::minutes(61));
     };
 
     // Run git fetch for updated repos
@@ -98,7 +98,7 @@ pub async fn get_changed_projects_since(
             match last_fetched {
                 Some(last_fetched)
                     if project
-                        .updated_at
+                        .last_activity_at
                         .as_ref()
                         .ok_or_else(|| eyre!("Missing update date for projects"))?
                         .0
-- 
GitLab


